crypto/src/merkle/merkle_tree.rs

```diff
@@ -0,0 +1,261 @@
+    ///
+    /// Merkle tree of depth 1 has two leaves, depth 2 has four leaves etc.
+    pub fn depth(&self) -> u32 {
+        log2(self.nodes.len() / 2)
+    }
+
+    /// Returns a node at the specified depth and index.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified depth is greater than the depth of the tree.
+    /// * The specified index not valid for the specified depth.
+    pub fn get_node(&self, depth: u32, index: u64) -> Result<Word, MerkleError> {
+        if depth == 0 {
+            return Err(MerkleError::DepthTooSmall);
+        } else if depth > self.depth() {
+            return Err(MerkleError::DepthTooBig(depth));
+        }
+        if index >= 2u64.pow(depth) {
+            return Err(MerkleError::InvalidIndex(depth, index));
+        }
+
+        let pos = 2usize.pow(depth as u32) + (index as usize);
+        Ok(self.nodes[pos])
+    }
+
+    /// Returns a Merkle path to the node at the specified depth and index. The node itself is
+    /// not included in the path.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified depth is greater than the depth of the tree.
+    /// * The specified index not valid for the specified depth.
+    pub fn get_path(&self, depth: u32, index: u64) -> Result<Vec<Word>, MerkleError> {
+        if depth == 0 {
+            return Err(MerkleError::DepthTooSmall);
+        } else if depth > self.depth() {
+            return Err(MerkleError::DepthTooBig(depth));
+        }
+        if index >= 2u64.pow(depth) {
+            return Err(MerkleError::InvalidIndex(depth, index));
+        }
+
+        let mut path = Vec::with_capacity(depth as usize);
+        let mut pos = 2usize.pow(depth as u32) + (index as usize);
+
+        while pos > 1 {
+            path.push(self.nodes[pos ^ 1]);
+            pos >>= 1;
+        }
+
+        Ok(path)
+    }
+
+    /// Replaces the leaf at the specified index with the provided value.
+    ///
+    /// # Errors
+    /// Returns an error if the specified index is not a valid leaf index for this tree.
+    pub fn update_leaf(&mut self, index: u64, value: Word) -> Result<(), MerkleError> {
+        let depth = self.depth();
+        if index >= 2u64.pow(depth) {
+            return Err(MerkleError::InvalidIndex(depth, index));
+        }
+
+        let mut index = 2usize.pow(depth) + index as usize;
+        self.nodes[index] = value;
+
+        let n = self.nodes.len() / 2;
+        let two_nodes =
+            unsafe { slice::from_raw_parts(self.nodes.as_ptr() as *const [Digest; 2], n) };
+
+        for _ in 0..depth {
+            index /= 2;
+            self.nodes[index] = merge(&two_nodes[index]).into();
+        }
+
+        Ok(())
+    }
+}
+
+// TESTS
+// ================================================================================================
+
+#[cfg(test)]
+mod tests {
+    use super::{Felt, FieldElement, Word};
+    use winterfell::crypto::{hashers::Rp64_256, ElementHasher, Hasher};
```

Comment:
Should we import the hasher from the `hash` module (instead of from Winterfell)?

Reply:
Done!